x-pack/plugins/task_manager/task_runner.ts

```diff
@@ -0,0 +1,258 @@
+   */
+  public toString() {
+    return `${this.taskType} "${this.id}"`;
+  }
+
+  /**
+   * Runs the task, handling the task result, errors, etc, rescheduling if need
+   * be. NOTE: the time of applying the middleware's beforeRun is incorporated
+   * into the total timeout time the task in configured with. We may decide to
+   * start the timer after beforeRun resolves
+   *
+   * @returns {Promise<RunResult>}
+   */
+  public async run(): Promise<RunResult> {
+    try {
+      this.logger.debug(`Running task ${this}`);
+      const modifiedContext = await this.beforeRun({
+        kbnServer: this.kbnServer,
+        taskInstance: this.instance,
+      });
+      const task = this.definition.createTaskRunner(modifiedContext);
+      this.task = task;
+      return this.processResult(this.validateResult(await this.task.run()));
+    } catch (error) {
+      this.logger.warning(`Task ${this} failed ${error.stack}`);
+      this.logger.debug(`Task ${JSON.stringify(this.instance)} failed ${error.stack}`);
+
+      return this.processResult({ error });
+    }
+  }
+
+  /**
+   * Attempts to claim exclusive rights to run the task. If the attempt fails
+   * with a 409 (http conflict), we assume another Kibana instance beat us to the punch.
+   *
+   * @returns {Promise<boolean>}
+   */
+  public async claimOwnership(): Promise<boolean> {
+    const VERSION_CONFLICT_STATUS = 409;
+
+    try {
+      this.instance = await this.store.update({
+        ...this.instance,
+        status: 'running',
+        runAt: intervalFromNow(this.definition.timeOut)!,
+      });
+
+      return true;
+    } catch (error) {
+      if (error.statusCode !== VERSION_CONFLICT_STATUS) {
+        throw error;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Attempts to cancel the task.
+   *
+   * @returns {Promise<void>}
+   */
+  public async cancel() {
+    const { task } = this;
+    if (task && task.cancel) {
+      this.task = undefined;
+      return task.cancel();
+    }
+
+    this.logger.warning(`The task ${this} is not cancellable.`);
+  }
+
+  private validateResult(result?: RunResult | void): RunResult {
+    const { error } = Joi.validate(result, validateRunResult);
+
+    if (error) {
+      this.logger.warning(`Invalid task result for ${this}: ${error.message}`);
+    }
+
+    return result || {};
+  }
+
+  private async processResult(result: RunResult): Promise<RunResult> {
+    const recurring = result.runAt || this.instance.interval || result.error;
+    if (recurring) {
+      // recurring task: update the task instance
+      const state = result.state || this.instance.state || {};
+      const status = this.instance.attempts < this.store.maxAttempts ? 'idle' : 'failed';
+
+      let runAt;
+      if (status === 'failed') {
+        // task run errored, keep the same runAt
+        runAt = this.instance.runAt;
+      } else {
+        runAt =
+          result.runAt ||
+          intervalFromNow(this.instance.interval) ||
```

Comment:
This is implicitly causing a subtle drift of running. As an example:

- Run at `00:00:00.000`, with an interval of `1m`.
- Running takes `200ms`.
- Next run won't be until `00:01:00.200`.

For v1, I think it's fine, but to fix it we will need more than just the `RunResult` -- we'll need the details about when we actually _tried_ to run it so that we can properly shift it.